File: Stack/carFleet.cpp
#include <bits/stdc++.h>
using namespace std;

int carFleet(int target, vector<int> &position, vector<int> &speed)
{
    deque<pair<int, double>> cars;

    for (int i = 0; i < position.size(); i++){
        double time = (double) (target - position[i]) / speed[i];


        cars.push_back({position[i], time});
    }

    sort(cars.begin(), cars.end());

    double maxTime = 0.0;
    int result = 0;
    for (int i = cars.size() -1; i>=0; i--){
        double time = cars[i].second;

        if (time > maxTime){
            maxTime = time;
            result ++;
        }



    }

    return result;
}

File: Stack/Car_Fleet.py
from typing import List


class Solution:
    def carFleet(self, target: int, position: List[int], speed: List[int]) -> int:
        pairs = []  # (pos, time)
        for i in range(len(position)):
            pairs.append((position[i], (target - position[i]) / speed[i]))

        pairs.sort()

        maxTime = 0.0
        res = 0

        for i in range(len(pairs) - 1, -1, -1):
            time = pairs[i][1]

            if time > maxTime:
                res += 1
                maxTime = time

        return res
